src/core/bpf-lsm.c

```diff
@@ -0,0 +1,404 @@
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to get cgroup ID for path '%s': %m", path);
+
+        inner_map_fd = sym_bpf_create_map(
+                        BPF_MAP_TYPE_HASH,
+                        sizeof(uint32_t),
+                        sizeof(uint32_t),
+                        128, /* Should be enough for all filesystem types */
+                        0);
+        if (inner_map_fd < 0)
+                return log_unit_error_errno(u, errno, "Failed to create inner LSM map: %m");
+
+        outer_map_fd = sym_bpf_map__fd(u->manager->restrict_fs->maps.cgroup_hash);
+        if (outer_map_fd < 0)
+                return log_unit_error_errno(u, errno, "Failed to get BPF map fd: %m");
+
+        if (sym_bpf_map_update_elem(outer_map_fd, &cgroup_id, &inner_map_fd, BPF_ANY) != 0)
+                return log_unit_error_errno(u, errno, "Error populating LSM BPF map: %m");
+
+        uint32_t allow = allow_list;
+
+        /* Use key 0 to store whether this is an allow list or a deny list */
+        if (sym_bpf_map_update_elem(inner_map_fd, &zero, &allow, BPF_ANY) != 0)
+                return log_unit_error_errno(u, errno, "Error initializing BPF map: %m");
+
+        SET_FOREACH(fs, filesystems) {
+                r = fs_type_from_string(fs, &magic);
+                if (r < 0) {
+                        log_unit_warning(u, "Invalid filesystem name '%s', ignoring.", fs);
+                        continue;
+                }
+
+                log_unit_debug(u, "Restricting filesystem access to '%s'", fs);
+
+                for (int i = 0; i < FILESYSTEM_MAGIC_MAX; i++) {
+                        if (magic[i] == 0)
+                                break;
+
+                        if (sym_bpf_map_update_elem(inner_map_fd, &magic[i], &dummy_value, BPF_ANY) != 0) {
+                                r = log_unit_error_errno(u, errno, "Failed to update BPF map: %m");
+
+                                if (sym_bpf_map_delete_elem(outer_map_fd, &cgroup_id) != 0)
+                                        log_unit_debug_errno(u, errno, "Failed to delete cgroup entry from LSM BPF map: %m");
+
+                                return r;
+                        }
+                }
+        }
+
+        return 0;
+}
+
+int cleanup_lsm_bpf(const Unit *u) {
+        _cleanup_free_ char *path = NULL;
+        uint64_t cgroup_id;
+        int fd = -1;
+        int r;
+
+        assert(u);
+        assert(u->manager);
+
+        if (!lsm_bpf_supported())
+                return 0;
+
+        if (!u->manager->restrict_fs)
+                return 0;
+
+        r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, NULL, &path);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to get cgroup path: %m");
+
+        r = cg_path_get_cgroupid(path, &cgroup_id);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to get cgroup ID: %m");
```

Comment:
I'm seeing a bunch of messages with this error:

```
[vagrant@archlinux ~]$ sudo journalctl -b -3 | grep "Failed to get cgroup ID"
Aug 19 14:54:56 archlinux systemd[1]: systemd-homed-activate.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:06 archlinux systemd[1]: systemd-coredump.socket: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:06 archlinux systemd[1]: git.mount: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:06 archlinux systemd[1]: vagrant.mount: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: run-user-1000.mount: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: dbus.socket: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-networkd-wait-online.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-update-done.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: ldconfig.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-hwdb-update.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-journal-catalog-update.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-tmpfiles-setup.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-networkd.socket: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-sysctl.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-tmpfiles-setup-dev.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-sysusers.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-remount-fs.service: Failed to get cgroup ID: No such file or directory
Aug 19 15:31:07 archlinux systemd[1]: systemd-reboot.service: Failed to get cgroup ID: No such file or directory
```

It seems name_to_handle_at() fails with ENOENT so the cgroup path might already be gone. I wonder if we should make this fail silently in this case.

Reply:
Maybe a better way is to save the `cgroup_id` in `struct Unit` next to the `cgroup_path` and then this can be used by the `bpf_restrict_filesystems` and `cleanup_lsm_bpf()`. Wdyt?

Reply:
Added a couple of commits implementing this. I can squash the bpf-lsm commit if this approach is ok.